pkg/keyctl/keyring.go

```diff
@@ -77,3 +82,68 @@ func Link(parent Keyring, child ID) error {
 	_, err := unix.KeyctlInt(unix.KEYCTL_LINK, int(child.ID()), int(parent.ID()), 0, 0)
 	return err
 }
+
+// ReadUserKeyring reads user keyring and returns slice of Key with id field(key_serial_t) representing the IDs of all the keys that are linked to it
+func (kr *keyring) ReadUserKeyring() ([]*Key, error) {
+	var (
+		b        []byte
+		err      error
+		sizeRead int
+		keyArr   []*Key
+	)
+
+	if kr.size == 0 {
+		kr.size = 4
+	}
+
+	size := kr.size
+
+	b = make([]byte, size)
+	sizeRead = size + 1
+	for sizeRead > size {
+		r1, err := unix.KeyctlBuffer(unix.KEYCTL_READ, unix.KEY_SPEC_USER_KEYRING, b, size)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if sizeRead = int(r1); sizeRead > size {
+			b = make([]byte, sizeRead)
+			size = sizeRead
+			sizeRead = size + 1
+		} else {
+			kr.size = sizeRead
+		}
+	}
+
+	keyIDs := getKeyIDsFromByte(b[:kr.size])
+	for _, kid := range keyIDs {
+		keyArr = append(keyArr, &Key{id: kid})
+	}
+	return keyArr, err
+}
+
+func getKeyIDsFromByte(byteKeyIDs []byte) []keyID {
+	idSize := 4
+
+	// check the endianness
+	var nativeEndian binary.ByteOrder
+	buf := [2]byte{}
+	*(*uint16)(unsafe.Pointer(&buf[0])) = uint16(0xABCD)
+
+	switch buf {
+	case [2]byte{0xCD, 0xAB}:
+		nativeEndian = binary.LittleEndian
+	case [2]byte{0xAB, 0xCD}:
+		nativeEndian = binary.BigEndian
+	default:
+		panic("Could not determine native endianness.")
+	}
```

Comment:
Ouch…

I mean, if this is going to be using `unsafe` at all, I guess it could just as well use something vaguely like (completely untested!)
```go
tempId := *(*int32)(unsafe.Pointer(&buf[idx:idx+idSize])
```
below, without bothering with the `encoding/binary` package at all, but finding a way to not use `unsafe` would be much nicer. Not that I know how to do that right now.

Reply:
I can use `*(*int32)(unsafe.Pointer(&byteKeyIDs[idx]))`